source/server/listener_manager_impl.cc

```diff
@@ -108,33 +109,67 @@ std::vector<Network::FilterFactoryCb> ProdListenerComponentFactory::createNetwor
+ProdListenerComponentFactory::createListenerFilterFactoryListImpl(
     const Protobuf::RepeatedPtrField<envoy::config::listener::v3::ListenerFilter>& filters,
-    Configuration::ListenerFactoryContext& context) {
-  std::vector<Network::ListenerFilterFactoryCb> ret;
+    Configuration::ListenerFactoryContext& context,
+    Filter::TcpListenerFilterConfigProviderManagerImpl& config_provider_manager) {
+  Filter::ListenerFilterFactoriesList ret;
+
+  ret.reserve(filters.size());
   for (ssize_t i = 0; i < filters.size(); i++) {
     const auto& proto_config = filters[i];
     ENVOY_LOG(debug, "  filter #{}:", i);
     ENVOY_LOG(debug, "    name: {}", proto_config.name());
     ENVOY_LOG(debug, "  config: {}",
               MessageUtil::getJsonStringFromMessageOrError(
                   static_cast<const Protobuf::Message&>(proto_config.typed_config())));
-
-    // Now see if there is a factory that will accept the config.
-    auto& factory =
-        Config::Utility::getAndCheckFactory<Configuration::NamedListenerFilterConfigFactory>(
-            proto_config);
-    auto message = Config::Utility::translateToFactoryConfig(
-        proto_config, context.messageValidationVisitor(), factory);
-    ret.push_back(factory.createListenerFilterFactoryFromProto(
-        *message, createListenerFilterMatcher(proto_config), context));
+    // dynamic listener filter configuration
+    if (proto_config.config_type_case() ==
+        envoy::config::listener::v3::ListenerFilter::ConfigTypeCase::kConfigDiscovery) {
+      const auto& config_discovery = proto_config.config_discovery();
+      const auto& name = proto_config.name();
+      ENVOY_LOG(debug, "   Listener filter:   dynamic filter name: {}", name);
+      if (config_discovery.apply_default_config_without_warming() &&
+          !config_discovery.has_default_config()) {
+        throw EnvoyException(fmt::format(
+            "Error: listener filter config {} applied without warming but has no default config.",
+            name));
+      }
+      for (const auto& type_url : config_discovery.type_urls()) {
+        const auto factory_type_url = TypeUtil::typeUrlToDescriptorFullName(type_url);
+        const auto* factory =
+            Registry::FactoryRegistry<Server::Configuration::NamedListenerFilterConfigFactory>::
+                getFactoryByType(factory_type_url);
+        if (factory == nullptr) {
+          throw EnvoyException(fmt::format(
+              "Error: no listener factory found for a required type URL {}.", factory_type_url));
+        }
+      }
+      auto filter_config_provider = config_provider_manager.createDynamicFilterConfigProvider(
+          config_discovery, name, context, "tcp_listener.", false, "listener",
+          createListenerFilterMatcher(proto_config));
+      ret.push_back(std::move(filter_config_provider));
+    } else {
+      // For static configuration, now see if there is a factory that will accept the config.
+      auto& factory =
```

Comment:
Comparing this with the [HttpFilter](https://github.com/envoyproxy/envoy/blob/6d85d09c7dbc7c3a86dddbe89a35c54f7c431cac/source/extensions/filters/network/http_connection_manager/config.cc#L638), it seems that Listener filters are required, whereas HttpFilter is optional.
I think it may make sense to add this capability, as it allows both Envoy that are aware of a new filter run with an Envoy that isn't aware of the filter (while receiving the same config). WDYT?

Reply:
The HttpFilter is_optional feature is added by the PR:https://github.com/envoyproxy/envoy/pull/16119.
It's a nice to have feature. However, it's not trivial to support it in ListenerFilter. Also it's orthogonal to the ECDS functionality this PR is trying to address.  Thus I would think this should be tracked by a separate issue.